src/combinatorics/simple_graph/metric.lean

```diff
@@ -0,0 +1,110 @@
+/-
+Copyright (c) 2022 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller, Vincent Beffara
+-/
+import combinatorics.simple_graph.connectivity
+import data.nat.lattice
+
+/-!
+# Graph metric
+
+This module defines the `simple_graph.dist` function, which takes
+pairs of vertices to the length of the shortest walk between them.
+
+## Main definitions
+
+- `simple_graph.dist` is the graph metric.
+
+## Todo
+
+- Provide an additional computable version of `simple_graph.dist`
+  for when `G` is connected.
+
+- Evaluate `nat` vs `enat` for the codomain of `dist`, or potentially
+  having an additional `edist` when the objects under consideration are
+  disconnected graphs.
+
+- When directed graphs exist, a directed notion of distance,
+  likely `enat`-valued.
+
+## Tags
+
+graph metric
+
+-/
+
+namespace simple_graph
+variables {V : Type*} (G : simple_graph V)
+
+/-! ## Metric -/
+
+/-- The distance between two vertices is the length of the shortest walk between them.
+If no such walk exists, this uses the junk value of `0`. -/
+noncomputable
+def dist (u v : V) : ℕ := Inf (set.range (walk.length : G.walk u v → ℕ))
+
+variables {G}
+
+protected
+lemma reachable.exists_walk_of_dist {u v : V} (hr : G.reachable u v) :
+  ∃ (p : G.walk u v), p.length = G.dist u v :=
+nat.Inf_mem (set.range_nonempty_iff_nonempty.mpr hr)
+
+protected
+lemma connected.exists_walk_of_dist (hconn : G.connected) (u v : V) :
+  ∃ (p : G.walk u v), p.length = G.dist u v :=
+(hconn u v).exists_walk_of_dist
+
+lemma dist_le {u v : V} (p : G.walk u v) : G.dist u v ≤ p.length := nat.Inf_le ⟨p, rfl⟩
+
+@[simp] lemma dist_self {v : V} : dist G v v = 0 :=
+le_antisymm (dist_le (walk.nil : walk G v v)) (zero_le _)
```

Comment:
Is it any easier to prove this with [nat.Inf_eq_zero](https://leanprover-community.github.io/mathlib_docs/data/nat/lattice.html#nat.Inf_eq_zero)?

Reply:
In fact, if you reorder the lemmas, the proof is `by simp`, so it doesn't need to be a simp lemma anymore.